java/daikon/dcomp/BuildJDK.java

```diff
@@ -3,375 +3,461 @@
+   * directory tree, selects the classes we want to instrument, creates an InputStream for each of
+   * these classes, and adds this information to the {@code class_stream_map} argument.
+   *
+   * @param path module file, which might be subdirectory
+   * @param modulePrefixLength length of "/module/..." path prefix before start of actual member
+   *     path
+   * @param class_stream_map a map from class file name to InputStream that collects the results
+   */
+  void gather_runtime_from_modules_directory(
+      Path path, int modulePrefixLength, Map<String, InputStream> class_stream_map) {
+
+    if (Files.isDirectory(path)) {
+      try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(path)) {
+        for (Path subpath : directoryStream) {
+          gather_runtime_from_modules_directory(subpath, modulePrefixLength, class_stream_map);
+        }
+      } catch (IOException e) {
+        throw new Error(e);
+      }
+    } else {
+      String entryName = path.toString().substring(modulePrefixLength + 1);
+      try {
+        // Get the InputStream for this file
+        InputStream is = Files.newInputStream(path);
+        class_stream_map.put(entryName, is);
+      } catch (Exception e) {
+        throw new Error(e);
+      }
+    }
+  }
+
+  /**
+   * Instrument each of the classes indentified by the class_stream_map argument.
+   *
+   * @param dest_dir where to store the instrumented classes
+   * @param class_stream_map maps from class file name to an input stream on that file
+   */
+  void instrument_classes(File dest_dir, Map<String, InputStream> class_stream_map) {
+
+    try {
+      // Create the destination directory
+      dest_dir.mkdirs();
+
+      // Process each file.
+      for (String classFileName : class_stream_map.keySet()) {
+        if (verbose) {
+          System.out.println("instrument_classes: " + classFileName);
+        }
+
+        if (classFileName.equals("module-info.class")) {
+          System.out.printf("Skipping file %s%n", classFileName);
           continue;
         }
-        if (entryName.endsWith(".class")) all_classes.add(entryName.replace(".class", ""));
-        if (!entryName.endsWith(".class")
-            || (skip_object && entryName.equals("java/lang/Object.class"))
-            || (!prefix.equals("") && !entryName.startsWith(prefix))) {
-          File destfile = new File(entryName);
-          if (destfile.getParent() == null) {
-            System.out.printf("Skipping file %s%n", destfile);
+
+        if (!classFileName.endsWith(".class") || classFileName.equals("java/lang/Object.class")) {
+          File destfile = new File(classFileName);
+          // Copy non-.class files (and Object.class) unchanged (JDK 8).
+          // For JDK 9+ we do not copy as these items will be loaded from the original module file.
+          if (destfile.getParent() == null || BcelUtil.javaVersion > 8) {
```

Comment:
When is `destfile.getParent()` null?  The comments don't say anything about that part of the test.

Reply:
test is to skip any non ".class" file at root of jar entries.  In a normal rt.jar file there are none.  Don't know why Ben put this there.  Though it does protect the use in the 'new File" call just below the end of the if.

Comment or remove it?  I lean towards remove it.  Its inside a try-catch that will report error if not a well formed rt.jar.

Reply:
The `if (!classFileName.endsWith(".class")` test just above skips non-`.class` files.
This test skips **`.class`** files at the root, I think.
I made changes in fd0bc8e72; could you review that commit?